### PR TITLE
composition: implement field type composition

### PR DESCRIPTION
**Pull request by tomhoule**
In input (argument, input object) and output (interface, object) fields.

closes GB-5269
closes GB-5176

### CODE REVIEW

File: engine/crates/composition/src/compose/fields.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        .try_fold((first, first.r#type()), |(a_arg, a_type), (b_arg, b_type)| {
+            a_type
+                .compose_for_input(b_type)
+                .map(|ty| (a_arg, ty))
+                .ok_or((a_arg, b_arg))
+        }) {
+        Ok((_, ty)) => Some(ty.id),
+        Err((a_arg, b_arg)) => {
+            ctx.diagnostics.push_fatal(format!(
+                "The `{}.{}({}:)` argument has conflicting types in different subgraphs: {} in {} but {} in {}",
```
Comment (tomhoule) [inline engine/crates/composition/src/compose/fields.rs:69]: remove the backticks for consistency

File: engine/crates/composition/src/subgraphs/field_types.rs [hunk 2]
```diff
@@ -84,6 +84,44 @@ impl<'a> FieldTypeWalker<'a> {
...
+    pub(crate) fn compose_for_input(self, other: Self) -> Option<Self> {
+        Some(if self.compose(other)? { other } else { self })
+    }
+
+    /// Compose two field types for output. The less required of the two is picked.
+    pub(crate) fn compose_for_output(self, other: Self) -> Option<Self> {
+        Some(if self.compose(other)? { self } else { other })
+    }
+
+    /// Returns if `other` is nonnullable. This is enough to make a decision about which to pick
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/field_types.rs:97]: returns whether

Comment (tomhoule) [reply engine/crates/composition/src/subgraphs/field_types.rs:97]: Explain the none case.

File: engine/crates/composition/tests/composition/field_type_mismatch/federated.graphql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# The `LettuceInterface.variety` field has conflicting types in different subgraphs: String! in lettuces_a but Int! in lettuces_b
+# The `Lettuce.variety` field has conflicting types in different subgraphs: String! in lettuces_a but Int! in lettuces_b
+# The `Lettuce.origin` field has conflicting types in different subgraphs: Int! in lettuces_a but String! in lettuces_b
+# The `Lettuce.leaves(onlyEdible:)` argument has conflicting types in different subgraphs: Int in lettuces_a but Boolean in lettuces_b
+# The `LettuceInput.variety` field has conflicting types in different subgraphs: String! in lettuces_a but Int! in lettuces_b
```
Comment (tomhoule) [inline engine/crates/composition/tests/composition/field_type_mismatch/federated.graphql:5]: TODO: add a case with a list type of different nullability

File: engine/crates/composition/tests/composition/field_type_composition/federated.graphql [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+interface Spinner {
+    id: ID!
+    model: String
+}
+
+input SpinnerFilter {
+    manufacturer: String
+    maxPrice: Float!
+}
```
Comment (tomhoule) [inline engine/crates/composition/tests/composition/field_type_composition/federated.graphql:46]: todo: add a list field with different nullability

File: engine/crates/composition/src/compose/input_object.rs [hunk 5]
```diff
@@ -7,53 +7,63 @@ pub(super) fn merge_input_object_definitions(
...
-            ));
-        }
-    }
+    all_fields.sort_by_key(|(name, _)| *name);
 
-    let composed_directives = collect_composed_directives(definitions.iter().map(|def| def.directives()), ctx);
+    let mut start = 0;
 
-    ctx.insert_input_object(first.name(), description, composed_directives);
+    while start < all_fields.len() {
```
Comment (tomhoule) [inline engine/crates/composition/src/compose/input_object.rs:31]: I want to move all these blocks into separate functions, but let's do that in a separate PR to avoid noise in this one.

File: engine/crates/composition/src/subgraphs/field_types.rs [hunk 6]
```diff
@@ -45,7 +45,7 @@ impl Subgraphs {
             match &ty.base {
                 ast::BaseType::List(inner) => {
                     let wrapper = WrapperType {
-                        kind: if inner.nullable {
+                        kind: if ty.nullable {
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/field_types.rs:48]: This is the list type handling bug.

File: engine/crates/composition/src/subgraphs/field_types.rs [hunk 7]
```diff
@@ -84,6 +84,58 @@ impl<'a> FieldTypeWalker<'a> {
             .map(|id| self.walk(id))
     }
 
+    /// Compose two field types for output. The most required of the two is picked.
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/field_types.rs:87]: typo: should say "input"

File: engine/crates/composition/src/subgraphs/field_types.rs [hunk 8]
```diff
@@ -84,6 +84,58 @@ impl<'a> FieldTypeWalker<'a> {
...
+    pub(crate) fn compose_for_input(self, other: Self) -> Option<Self> {
+        Some(if self.compose(other)? { other } else { self })
+    }
+
+    /// Compose two field types for output. The less required of the two is picked.
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/field_types.rs:92]: s/less/least

File: engine/crates/composition/src/subgraphs/field_types.rs [hunk 9]
```diff
@@ -120,3 +172,22 @@ impl<'a> FieldTypeWalker<'a> {
             .unwrap_or_else(|| self.inner().is_required)
     }
 }
+
+impl std::fmt::Display for FieldTypeWalker<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut out = self.type_name().as_str().to_owned();
```
Comment (tomhoule) [inline engine/crates/composition/src/subgraphs/field_types.rs:178]: Lots of allocations, but it's only for errors so I don't think we should care, it doesn't slow down the happy path.

**Comment by tomhoule** [comment]
Some snapshots need updating in integration tests.

**Review by obmarg — APPROVED** [review]
